### ISSUE TITLE
Java 22 Support

### ISSUE THREAD

**Issue by tugalsan** [yWorks/yGuard#153]
Will there be incremental support, like Java 22; or will wait for LTS, like Java 25?

**Comment by thomasbehr** [comment]
We have not decided yet.

**Comment by yGuy** [comment]
For some releases Oracle decided to massively change the bytecode and thus supporting the new versions was a lot of effort. Then again for other releases the bytecode doesn't change at all and supporting the new version is mostly a one-line change.

Someone will need to take a look at the bytecode changes so that we can estimate the time it requires to update yGuard accordingly. It all depends...

**Comment by Fohlen** [comment]
Since Java 21 is the LTS release of Java for the foreseeable future (7 years), Java 22 support is not high on the priority list. However @tugalsan you can always kickstart this process. The first step is to research the Java bytecode instruction set. You can diff Java 21 and 21 [from here](https://docs.oracle.com/javase/specs/index.html). Once you have a list of new bytecode instructions compiled, we can decide which ones need adjustments. If there are no breaking changes, we can get this rolling quickly. Would you be willing to compile such a list for us? 😉

**Comment by tugalsan** [comment]
@Fohlen Did i understand correctly? As a first step, https://docs.oracle.com/javase/specs/jvms/seXXX/html/index.html site needs to be compared.

I created a HelloWorld project with Java22 and it worked. I am so confused.

Which version X should be compared against 22? What is the latest supported Java of YGuard?

**Comment by thomasbehr** [comment]
yGuard supports Java 21.

I think Fohlen meant to suggest comparing the [Java 21 Virtual Machine Specification](https://docs.oracle.com/javase/specs/jvms/se21/html/index.html) to the [Java 22 Virtual Machine Specification](https://docs.oracle.com/javase/specs/jvms/se22/html/index.html).

Your HelloWorld project probably works, because your source code is essentially Java 8. I am not completely sure, but I think the Java compiler will actually _not_ emit Java 22 byte code for such simple cases but emit Java 8 byte code. You probably need to add Java 22 language features for force the compiler to emit Java 22 byte code. (You can use javap to check the class file versions of the emitted byte code.)

**Comment by tugalsan** [comment]
Yea, pom has a problem, it was 21 eventually, my bad.

**Comment by Benetl2007** [comment]
hello guys,
Thanks for the release of yGuard **5.0.0**.

I did some tests about this release, it does the obfuscation of the code with JDK 25 (as expected).
I was forced to do some changes in the yguard maven-antrun-plugin configuration of my Java project:
- replace `<inoutpairs resources="auto">` by `<inoutpairs> (blocking otherwise)`
- remove `<shrink logfile="target/yguard.log.xml"> part (blocking otherwise)`

The diff between my target/yguard.log.xml 5.0.0 and the one on 4.1.1 displays surprisingly lots of changes.
Some map values for field are looking like "map=`"&#x00a3;"`, it was map=[a-z;A-Z;_]+ as far (I remember).

The doc setup page still speaks about version 4.1.1: [https://github.com/yWorks/yGuard/blob/master/docs/setup.md
](url)
Thanks for your work,

Regards,

**Comment by Fohlen** [comment]
Hi @Benetl2007  could you attach the diff you are talking about?

In my opinion the config is largely the same, except for the fact that the shrinker has been removed, and cannot be configured any longer.

**Comment by Benetl2007** [comment]
hello @Fohlen,
here are the 4.1.1 and 5.0.0 yguard.log.xml files (replacing the name of my own application by myapp, package by mypackage, main class by MyMain but the rest of the files is original).
Many differences and special encoding characters for field map values.
Regards,

[yguard_5_0_0.log.xml](https://github.com/user-attachments/files/26164795/yguard_5_0_0.log.xml)
[yguard_4_1_1.log.xml](https://github.com/user-attachments/files/26164796/yguard_4_1_1.log.xml)

**Comment by thomasbehr** [comment]
@Benetl2007 As Lennard already mentioned, support for shrinking has been removed from yGuard with version 5.0.0. This means the `resources` attribute for `inoutpairs` and the `shrink` element do no longer exist / are no longer supported. As a result, if either the `resources` attribute or the `shrink` element is used in a yGuard task, ANT throws a BuildException - i.e.
```
inoutpairs doesn't support the "resources" attribute
```
or
```
yguard doesn't support the nested "shrink" element.
```

If you use yGuard from Maven, the ANT BuildException will result in a MojoExecutionException for Maven's AntRun plugin by default. If you use `<failOnError>false</failOnError>` for Maven's AntRun plugin, the Maven build continues but prints the ANT BuildException stacktrace. At least that is the behavior for me with Maven 3.9.14.

Finally, the differences in mapping results between 4.1.1 and 5.0.0 are to be expected, if you used `<shrink>` for 4.1.1. The reason for this is the way renaming works. Essentially, renaming traverses all types and type members and for each type/type member a name generator is queried for a "next" name. E.g.
```
com -> A
com.mycompany -> A
com.mycompany.Class1 -> A
com.mycompany.Class2 -> B
com.mycompany.Class2#method1 -> a
com.mycompany.Class2#method2 -> b
com.mycompany.Class3 -> C
```
Now, suppose `Class1` and `Class2#method1` are never used and thus `<shrink>` removes them. Then the renaming result for the above example would be
```
com -> A
com.mycompany -> A
com.mycompany.Class2 -> A
com.mycompany.Class2#method2 -> a
com.mycompany.Class3 -> B
```
In short, using `<shrink>` changes the result of `<rename>`. (Or in your case, not using `<shrink>` anymore changes the result of `<rename>`.)
